### PR TITLE
Adding Check abstraction and loop

### PR DESCRIPTION
**Pull request by nhibberd**
! @markhibberd @tmcgilchrist @jystic

### CODE REVIEW

File: src/Twine/Loop.hs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import           Twine.Data.Check
+import           Twine.Data.Duration
+import           Twine.Snooze
+
+
+-- | Loop with a delay until the continue check fails
+--
+loop :: MonadIO m => Duration -> Check -> m () -> m ()
+loop d c action = do
+  action
```
Comment (thumphries) [inline src/Twine/Loop.hs]: Should we care if `action` throws an exception? or assume it's been wrapped up in some handler already

**Comment by thumphries** [comment]
This construct looks threadsafe to me, appropriate use of IORef. 👍 

It looks a lot like `Pin`, which is also nonblocking. Also, `loop` could be written in terms of `Twine.Guard.guarded` if you added `Break` to the `TerminationAction` type, giving it explicit exception handling. Could you explain why Check and Loop should exist as new concepts?

**Comment by nhibberd** [comment]
> It looks a lot like Pin, which is also nonblocking.

`newPin >>= waitForPin` should be blocking which was the main reason for the new construct of `Check`

As for `guarded`, perhaps I should add something else to the comment but ideally loop is designed to be used inside an `Async` where it would be not safe to use `guarded`

File: src/Twine/Loop.hs (hunk 1 above)
Comment (tmcgilchrist) [inline src/Twine/Loop.hs]: Curious about this too. What happens under an exception?

**Comment by thumphries** [comment]
Just making sure I understood you earlier:
- Assume some exception handler is on the outside (like Async, which might try to kill the thread etc)
- `waitForPin` is blocking, and we want to prevent misuse at the API level. Cancel will have a different pattern of use to Pin, nobody cares that they're the same thing

Seems like you might want to encode the Async usage pattern here, since `loop` is unsafe.

🍙  as-is if you're happy with it

File: src/Twine/Loop.hs (hunk 1 above)
Comment (nhibberd) [inline src/Twine/Loop.hs]: This abstraction is not concerned with how they are handled, the current use case would be that the loop is wrapped in something that would handle them

**Comment by nhibberd** [comment]
> Assume some exception handler is on the outside (like Async, which might try to kill the thread etc)

+1

> waitForPin is blocking, and we want to prevent misuse at the API level. Cancel will have a different pattern of use to Pin, nobody cares that they're the same thing

+1

**Comment by nhibberd** [comment]
Updated names as per request @jystic

**Comment by tmcgilchrist** [comment]
🌵 🍨

**Comment by jystic** [comment]
💰

**Pull request merged**